network/p2p/unicast/manager.go

```diff
@@ -138,49 +231,90 @@ func (m *Manager) rawStreamWithProtocol(ctx context.Context,
-
 		err := m.streamFactory.Connect(ctx, peer.AddrInfo{ID: peerID})
 		if err != nil {
-
 			// if the connection was rejected due to invalid node id, skip the re-attempt
 			if strings.Contains(err.Error(), "failed to negotiate security protocol") {
-				return s, dialAddr, fmt.Errorf("invalid node id: %w", err)
+				return fmt.Errorf("invalid node id: %w", err)
 			}
 
 			// if the connection was rejected due to allowlisting, skip the re-attempt
 			if errors.Is(err, swarm.ErrGaterDisallowedConnection) {
-				return s, dialAddr, fmt.Errorf("target node is not on the approved list of nodes: %w", err)
+				return fmt.Errorf("target node is not on the approved list of nodes: %w", err)
 			}
-
+			m.logger.Warn().
+				Err(err).
+				Str("peer_id", peerID.String()).
+				Int64("attempt", dialAttempts.Load()).
+				Int("max_attempts", maxAttempts).
+				Msg("retrying peer dialing")
 			errs = multierror.Append(errs, err)
-			continue
+			return retry.RetryableError(errs)
+		}
+		return nil
+	}
+
+	// retryable func that will attempt to create the stream using the stream factory if connection exists
+	connectAttempts := atomic.NewInt64(0)
+	connectPeer := func(context.Context) error {
+		connectAttempts.Inc()
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context done before stream could be created (retry attempt: %s, errors: %w)", connectAttempts.String(), errs)
+		default:
 		}
 
+		var err error
+		// add libp2p context value NoDial to prevent the underlying host from dialingComplete the peer while creating the stream
+		// we've already ensured that a connection already exists.
+		ctx = libp2pnet.WithNoDial(ctx, "application ensured connection to peer exists")
 		// creates stream using stream factory
 		s, err = m.streamFactory.NewStream(ctx, peerID, protocolID)
 		if err != nil {
 			// if the stream creation failed due to invalid protocol id, skip the re-attempt
 			if strings.Contains(err.Error(), "protocol not supported") {
-				return nil, dialAddr, fmt.Errorf("remote node is running on a different spork: %w, protocol attempted: %s", err, protocolID)
+				return fmt.Errorf("remote node is running on a different spork: %w, protocol attempted: %s", err, protocolID)
 			}
 			errs = multierror.Append(errs, err)
-			continue
+			return retry.RetryableError(errs)
 		}
+		return nil
+	}
 
-		break
+	isConnected, err := m.connStatus.IsConnected(peerID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check connection status and attempt to dial the peer if dialing is not in progress
+	if !isConnected {
+		// return error if we can't start dialing
+		if m.dialingInProgress(peerID) {
+			return nil, nil, NewDialInProgressErr(peerID)
+		}
+		defer m.dialingComplete(peerID)
+		err = retry.Do(ctx, backoff, dialPeer)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
-	if retries == maxAttempts {
-		return s, dialAddr, errs
+	// at this point dialing should have completed we are already connected we can attempt to create the stream
```

Comment:
If all dial attempts fail IsConnected will never return true and subsequent calls to CreateStream would kick off dialing process, there is a retry on the initial call to CreateStream this will call IsConnected. I don't think we need to check it again at this point.